src/CodeGen_X86.cpp

```diff
@@ -360,61 +345,43 @@ void CodeGen_X86::visit(const Cast *op) {
         string intrin;
         Expr pattern;
     };
 
     // clang-format off
     static Pattern patterns[] = {
-        {true, "saturating_add", i8_sat(wild_i16x_ + wild_i16x_)},
-        {true, "saturating_sub", i8_sat(wild_i16x_ - wild_i16x_)},
-        {true, "saturating_add", i16_sat(wild_i32x_ + wild_i32x_)},
-        {true, "saturating_sub", i16_sat(wild_i32x_ - wild_i32x_)},
-        {true, "saturating_add", u8_sat(wild_u16x_ + wild_u16x_)},
-        {true, "saturating_sub", u8(max(wild_i16x_ - wild_i16x_, 0))},
-        {true, "saturating_add", u16_sat(wild_u32x_ + wild_u32x_)},
-        {true, "saturating_sub", u16(max(wild_i32x_ - wild_i32x_, 0))},
-
-        {true, "pmulh", i16((wild_i32x_ * wild_i32x_) / 65536)},
-        {true, "pmulh", u16((wild_u32x_ * wild_u32x_) / 65536)},
-        {true, "pmulhr", i16((((wild_i32x_ * wild_i32x_) + 16384)) / 32768)},
-
-        {true, "rounding_halving_add", u8(((wild_u16x_ + wild_u16x_) + 1) / 2)},
-        {true, "rounding_halving_add", u16(((wild_u32x_ + wild_u32x_) + 1) / 2)},
-
-        {false, "saturating_narrow", i16_sat(wild_i32x_)},
-        {false, "saturating_narrow", u16_sat(wild_i32x_)},
-        {false, "saturating_narrow", i8_sat(wild_i16x_)},
-        {false, "saturating_narrow", u8_sat(wild_i16x_)},
+        {"pmulh", i16(widening_mul(wild_i16x_, wild_i16x_) >> u32(16))},
+        {"pmulh", u16(widening_mul(wild_u16x_, wild_u16x_) >> u32(16))},
+        {"pmulhr", i16(rounding_shift_right(widening_mul(wild_i16x_, wild_i16x_), u32(15)))},
+
+        {"saturating_narrow", i16_sat(wild_i32x_)},
+        {"saturating_narrow", u16_sat(wild_i32x_)},
+        {"saturating_narrow", i8_sat(wild_i16x_)},
+        {"saturating_narrow", u8_sat(wild_i16x_)},
     };
     // clang-format on
 
+    vector<Expr> matches;
     for (size_t i = 0; i < sizeof(patterns) / sizeof(patterns[0]); i++) {
         const Pattern &pattern = patterns[i];
         if (expr_match(pattern.pattern, op, matches)) {
-            bool match = true;
-            if (pattern.wide_op) {
-                // Try to narrow the matches to the target type.
-                for (size_t i = 0; i < matches.size(); i++) {
-                    matches[i] = lossless_cast(op->type, matches[i]);
-                    if (!matches[i].defined()) {
-                        match = false;
-                    }
-                }
-            }
-            if (match) {
-                value = call_overloaded_intrin(op->type, pattern.intrin, matches);
-                if (value) {
-                    return;
-                }
+            value = call_overloaded_intrin(op->type, pattern.intrin, matches);
+            if (value) {
+                return;
             }
         }
     }
 
+    if (const Call *mul = Call::as_intrinsic(op->value, {Call::widening_mul})) {
+        if (op->value.type().bits() < op->type.bits() && op->type.bits() <= 32) {
+            // LLVM/x86 really doesn't like 8 -> 16 bit multiplication.
```

Comment:
This comment could use some more elaboration. The condition looks like it also applies to 16 -> 32 bit widening muls?